extensions/tunnel-forwarding/src/extension.ts

```diff
@@ -0,0 +1,293 @@
+		...args: unknown[]
+	) {
+		if (!this.outputChannel) {
+			this.outputChannel = vscode.window.createOutputChannel(this.label, { log: true });
+			vscode.commands.executeCommand('setContext', 'tunnelForwardingHasLog', true);
+		}
+		this.outputChannel[logLevel](message, ...args);
+	}
+}
+
+class TunnelProvider implements vscode.TunnelProvider {
+	private readonly tunnels = new Set<Tunnel>();
+	private readonly stateChange = new vscode.EventEmitter<StateT>();
+	private _state: StateT = { state: State.Inactive };
+
+	private get state(): StateT {
+		return this._state;
+	}
+
+	private set state(state: StateT) {
+		this._state = state;
+		this.stateChange.fire(state);
+	}
+
+	public readonly onDidStateChange = this.stateChange.event;
+
+	constructor(private readonly logger: Logger) { }
+
+	/** @inheritdoc */
+	public async provideTunnel(tunnelOptions: vscode.TunnelOptions): Promise<vscode.Tunnel> {
+		const tunnel = new Tunnel(
+			tunnelOptions.remoteAddress,
+			(tunnelOptions.privacy as TunnelPrivacyId) || TunnelPrivacyId.Private,
+		);
+
+		this.tunnels.add(tunnel);
+		tunnel.onDidDispose(() => {
+			this.tunnels.delete(tunnel);
+			this.updateActivePortsIfRunning();
+		});
+
+		switch (this.state.state) {
+			case State.Error:
+			case State.Inactive:
+				await this.setupPortForwardingProcess();
+			// fall through since state is now starting
+			case State.Starting:
+				this.updateActivePortsIfRunning();
+				return new Promise<Tunnel>((resolve, reject) => {
+					const l = this.stateChange.event(state => {
+						if (state.state === State.Active) {
+							tunnel.setPortFormat(state.portFormat);
+							l.dispose();
+							resolve(tunnel);
+						} else if (state.state === State.Error) {
+							l.dispose();
+							reject(new Error(state.error));
+						}
+					});
+				});
+			case State.Active:
+				tunnel.setPortFormat(this.state.portFormat);
+				this.updateActivePortsIfRunning();
+				return tunnel;
+		}
+	}
+
+	/** Re/starts the port forwarding system. */
+	public async restart() {
+		this.killRunningProcess();
+		await this.setupPortForwardingProcess(); // will show progress
+		this.updateActivePortsIfRunning();
+	}
+
+	private isInStateWithProcess(process: ChildProcessWithoutNullStreams) {
+		return (
+			(this.state.state === State.Starting || this.state.state === State.Active) &&
+			this.state.process === process
+		);
+	}
+
+	private killRunningProcess() {
+		if (this.state.state === State.Starting || this.state.state === State.Active) {
+			this.logger.log('info', '[forwarding] no more ports, stopping forwarding CLI');
+			this.state.process.kill();
+			this.state = { state: State.Inactive };
+		}
+	}
+
+	private updateActivePortsIfRunning() {
+		if (this.state.state !== State.Starting && this.state.state !== State.Active) {
+			return;
+		}
+
+		const ports = [...this.tunnels].map(t => ({ number: t.remoteAddress.port, privacy: t.privacy }));
+		this.state.process.stdin.write(`${JSON.stringify(ports)}\n`);
```

Comment:
Not sure if this is to update the state considering it's stdin

Reply:
Yea, the forwarding process reads from its stdin to take updates to the forwarded ports